### ISSUE TITLE
LH v7: Incorrect offline warning with passed Offline Audit

### ISSUE THREAD

**Issue by Joelsz** [GoogleChrome/lighthouse#11869]
#### Provide the steps to reproduce
1. Run LH on https://www.bhphotovideo.com

#### Issue 1
The PWA audit shows the following warning: "Warnings: Page does not work offline. The page will not be regarded as installable after Chrome 93, stable release August 2021."
While at the same time the Offline Audit "Registers a service worker that controls page and start_url" passes as expected.

#### What is the expected behavior?
If the Offline Audit passes there should be no warning about the page not working offline.

#### Issue 2
There's also another issue that the Audit for "Content is not sized correctly for the viewport" is failing, probably due to being calculated against the 1920px wide screen, while the emulated screen size is 360px wide.

#### What is the expected behavior?
"Content is not sized correctly for the viewport" Audit should be calculated against the emulated screen size and not the full screen size.

See screenshot below with both issues.

#### Environment Information
* Affected Channels: DevTools (haven't tried others)
* Lighthouse version: 7.0.0
* Chrome version: Version 89.0.4364.0 (Official Build) canary (64-bit)
* Node.js version:
* Operating System: Windows 10

![image](https://user-images.githubusercontent.com/6243173/102913861-630e8a80-444d-11eb-8868-7472504015ae.png)

**Comment by patrickhulce** [comment]
Thanks for filing @Joelsz! 

The same message appears in the Chrome console and application panel and appeared to be accurate from my attempt. When loading the site for the first time in a clean profile and then reloading while offline, I get the offline dino page. It's only after refreshing the page a couple times that the offline functionality appears to kick-in.

![image](https://user-images.githubusercontent.com/2301202/103556222-124c5800-4e77-11eb-9c13-d893022a9156.png)

**Comment by Joelsz** [comment]
Is this the expected behaviour?

**Comment by patrickhulce** [comment]
> Is this the expected behaviour? Does this mean that the Service Worker has an error?

Yes, I would expect this behavior from Lighthouse given how https://www.bhphotovideo.com behaves. If the intention was that this page should work offline after the first load, then yes the service worker has an error that does not accomplish this behavior. 

> In my opinion, the offline status for the different audits and for application tab should at least be consistent across all channels.

They are consistent, the DevTools application panel returns the exact same message on the first load like Lighthouse does (see above screenshot).

**Comment by Joelsz** [comment]
After investigating a bit, it seems to me that the Service Worker fails to install due to a Network Error caused by the "importScript" fetch (which is a third party domain). With that, I guess, it's on bhphotovideo.com to fix.

With not being consistent I meant the "Registers a service worker that controls the page" audit, which shouldn't pass in the case when the offline audit fails, since the service worker is not at all in control.

**Comment by patrickhulce** [comment]
> With not being consistent I meant the "Registers a service worker that controls the page" audit, which shouldn't pass in the case when the offline audit fails, since the service worker is not at all in control.

I see what you're saying. Ironically this is a bug *caused by consistency* with Chromium 😆 Chromium doesn't issue an installabillity error for the service worker failing to install, it only notices it's failed to install when it tests the offline functionality of the (non-existent) service worker.

**Comment by Joelsz** [comment]
> Chromium doesn't issue an installabillity error for the service worker failing to install, it only notices it's failed to install when it tests the offline functionality of the (non-existent) service worker.

I'm not sure about the infrastructure of Lighthouse, but the offline detection for the message and the offline detection for the "Registers a service worker that controls the page" audit, should come from the same source and should essentially be detected only once.

(The reason for me asking this, is because a failing audit would trigger a much prioritized fix than a message that is only a chrome update with new requirements)

**Comment by patrickhulce** [comment]
That is what will happen in August 2021. 

Today Chrome emits two things for the offline.

- The installability check that doesn't care if the service worker actually works. This passes in DevTools (you don't see it in the application pane) and passes in Lighthouse.
- The installability check that *does* care if the service worker actually works. This warns in DevTools and warns in Lighthouse.

The problem here is that the weaker check that doesn't catch this bug is currently used as the passing criteria in Chrome, and so by matching Chrome, Lighthouse cannot do anything other than warn here.

**Comment by Joelsz** [comment]
So I understand that after August 2021, the offline audit is going to fail in such a scenario.

Is there an issue where this change can be tracked?

**Comment by patrickhulce** [comment]
[This Chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=965802&q=installability%20error&can=2) appears to be where updates are being captured.

**Comment by Joelsz** [comment]
While debugging and trying to reproduce the issue, we found that running Lighthouse from DevTools always fails the offline audit with the following: `start_url does not respond with a 200 when offline. The start_url did respond, but not via a service worker`. (The "Current page responds with a 200 when offline" audit still passes)

However, running Lighthouse from command line with Puppeteer always passes the above audit.

Which one is the correct one and is this a concern?

(Using Chrome 88, Lighthouse 6.x and Puppeteer 6.x pointing to the same Chrome and Profile as DevTools.)

**Comment by patrickhulce** [comment]
> Which one is the correct one 

Impossible to say without knowing the entirety of the settings used for both runs and what situations the site wants to work offline. It's quite likely they're both "correct" but they're just testing different scenarios (e.g. one has a particular cache or service worker starting point and the other has a slightly different starting point), one of which the site supports offline and the other it doesn't.

> and is this a concern?

Probably, especially if you're still getting the warning in Lighthouse 7.0.

**Comment by Joelsz** [comment]
Thank you for your response @patrickhulce, yes, the message still shows on Lighthouse 7.

My assumption is that the `importScript` to a third party domain fails, and that's what causes the offline audit to fail, however, the SW seems to install just fine in a regular browser session.

1. Could that failure be because Lighthouse does not allow 3P domains for some reason in the context it tries to install the SW? Regardless if this is the case or not, it feels that this audit does not mimic a real user experience since this error is not reproducible in a simple browser session (unable to reproduce with your example above either)

2. How can I confirm that the 3P `importScript` is the cause of this failure or something else, are there any logs stored to look at?

Again, Thank you for your help!

**Comment by patrickhulce** [comment]
> Could that failure be because Lighthouse does not allow 3P domains for some reason in the context it tries to install the SW

Lighthouse doesn't do anything to the service worker. There shouldn't be any odd 3P limitations. I can also reproduce exactly what Lighthouse is complaining about as a regular user by...

1. Enabling Moto 3G emulation.
2. Navigation to https://bhphotovideo.com
3. Setting Network emulation to "Offline"
4. Reloading the page and getting the offline dino

Desktop version seems to work offline just fine (The "You're offline" page comes up), but that's not what Lighthouse is testing.

**Comment by Joelsz** [comment]
After some more digging... it seems that the check for `navigator.onLine` in the fetch handler is causing Lighthouse to fail the offline detection.

Removing that check and allowing all requests to continually be handled by the fetch handler, seems to have fixed the problem.

...but only in Lighthouse v7 (Chrome 90) because of #10237 , in v6 it still fails.

**Comment by Joelsz** [comment]
> After some more digging... it seems that the check for `navigator.onLine` in the fetch handler is causing Lighthouse to fail the offline detection.
> 
> Removing that check and allowing all requests to continually be handled by the fetch handler, seems to have fixed the problem.
> 
> ...but only in Lighthouse v7 (Chrome 90), probably because of #10237 , in v6 it still fails.

I mistakenly linked the wrong issue, #709 is the correct one.

**Comment by Joelsz** [comment]
@patrickhulce isn't this a Chrome bug that should be fixed by Chrome team? Here's the fetch handler that fails the offline detection from Chrome due to checking for `navigator.onLine`.

	async function swFetch(event){
		
		if(navigator.onLine || event.request.method != 'GET'){ 
			return;
		}
		
		event.respondWith(
			caches.match(event.request).then(function(response) {
				return response || fetch(event.request);
			}).catch(error => {
				return getOfflinePage();
			})
		);

	}
	
	self.addEventListener('fetch', e => e.waitUntil(swFetch(e)));

**Comment by paulirish** [comment]
I know there's been some more work on https://bugs.chromium.org/p/chromium/issues/detail?id=965802 so best to track that for this issue.
